src/huggingface_hub/_inference.py

```diff
@@ -0,0 +1,100 @@
+# - handle parameters? we can based implementation on inference.js
+# - validate inputs/options/parameters? with Pydantic for instance? or only optionally?
+# - add all tasks
+# - handle async requests
+
+
+class InferenceClient:
+    def __init__(
+        self, model: Optional[str] = None, token: Optional[str] = None, timeout: Optional[int] = None
+    ) -> None:
+        # If set, `model` can be either a repo_id on the Hub or an endpoint URL.
+        self.model: Optional[str] = model
+        self.headers = build_hf_headers(token=token)
+        self.timeout = timeout
+
+    def __repr__(self):
+        return f"<InferenceClient(model='{self.model if self.model else ''}', timeout={self.timeout})>"
+
+    def post(
+        self,
+        json: Optional[Union[str, Dict, List]] = None,
+        data: Optional[bytes] = None,
+        model: Optional[str] = None,
+        task: Optional[str] = None,
+    ) -> Response:
+        url = self._resolve_url(model, task)
+
+        if data is not None and json is not None:
+            warnings.warn("Ignoring `json` as `data` is passed as binary.")
+
+        response = get_session().post(url, json=json, data=data, headers=self.headers, timeout=self.timeout)
+        hf_raise_for_status(response)
+        return response
+
+    def summarization(
+        self,
+        text: str,
+        parameters: Optional[Dict[str, Any]] = None,
+        model: Optional[str] = None,
+    ):
+        payload: Dict[str, Any] = {"inputs": text}
+        if parameters is not None:
+            payload["parameters"] = parameters
+        response = self.post(json=payload, model=model, task="summarization")
+        return response.json()[0]["summary_text"]
+
+    def _resolve_url(self, model: Optional[str], task: Optional[str]) -> str:
+        model = model or self.model
+
+        # If model is already a URL, ignore `task` and return directly
+        if model is not None and (model.startswith("http://") or model.startswith("https://")):
+            return model
+
+        # # If no model but task is set => fetch the recommended one for this task
+        if model is None:
+            if task is None:
+                raise ValueError(
+                    "You must specify at least a model (repo_id or URL) or a task, either when instantiating"
+                    " `InferenceClient` or when making a request."
+                )
+            model = get_model_id_by_task(task)
+
+        # If no task but model is set => fetch the default task for this pipeline
+        if task is None:
+            task = get_task_by_model_id(model)
+
+        # Compute InferenceAPI url
+        return self.get_inference_api_url(model, task)
+
+    @staticmethod
+    def get_inference_api_url(model_id: str, task: str) -> str:
+        return f"{INFERENCE_ENDPOINT}/pipeline/{task}/{model_id}"
```

Comment:
Would like a confirmation on this to be sure but this is the URL we are already using in `huggingface_hub` and it is the case since [the first PR](https://github.com/huggingface/huggingface_hub/pull/65/files#diff-f365bea0eb892dc99a45fae684d3df7562fd42011be7dedb3e41e35fed94f298R117) (2y ago).

Reply:
Yeah :thinking: I think we never officially communicated about this endpoint nor documented it as we pushed people to use the official one, but this one is more flexible and I think is ok to use here

Reply:
Yes it's not officially documented as it causes confusion, and generally it's easier for everyone to keep 1 model - 1 task.

Biggest use case for this endpoint would be `feature-extraction` vs `sentence-similarity` (similarity is easier to understand in the widget form, but feature-extraction is the most useful one in a production setting where you're usually building a semantic database.

Reply:
Thanks for the details @Narsil. I think I'll keep the endpoint url with this form (i.e. /task/...). In the end it's just a url built for internal use so I can make the method private. Most users will not even realize it if they use the correct method for their model. 

And if a user tries to call a task on a model that doesn't support it, I'll gracefully handle the error to print to the user the available task(s) for their model.

Reply:
Realized after some experiments that if I can a wrong model <> task pair, the pipeline is loaded by Inference API and then an HTTP 500 is returned. Not so efficient to catch client-side (and server-side). Since this use case is mostly useful for feature-extraction task, I think I'll hardcode this one and use the more officially promoted `https://api-inference.huggingface.co/models/` URL in all other cases.